Mathlib/GroupTheory/CommutingProbability.lean

```diff
@@ -132,3 +136,86 @@ theorem inv_card_commutator_le_commProb : (↑(Nat.card (commutator G)))⁻¹ 
     (le_trans (ge_of_eq (commProb_eq_one_iff.mpr (Abelianization.commGroup G).mul_comm))
       (commutator G).commProb_quotient_le)
 #align inv_card_commutator_le_comm_prob inv_card_commutator_le_commProb
+
+-- Construction of group with commuting probability 1/n
+namespace DihedralGroup
+
+lemma commProb_odd {n : ℕ} (hn : Odd n) :
+    commProb (DihedralGroup n) = (n + 3) / (4 * n) := by
+  rw [commProb_def', DihedralGroup.card_conjClasses_odd hn, nat_card]
+  qify [show 2 ∣ n + 3 by rw [Nat.dvd_iff_mod_eq_zero, Nat.add_mod, Nat.odd_iff.mp hn]; rfl]
+  rw [div_div, ← mul_assoc]
+  congr
+
+private lemma div_two_lt {n : ℕ} (h0 : n ≠ 0) : n / 2 < n :=
+  Nat.div_lt_self (Nat.pos_of_ne_zero h0) (lt_add_one 1)
+
+private lemma div_four_lt : {n : ℕ} → (h0 : n ≠ 0) → (h1 : n ≠ 1) → n / 4 + 1 < n
+  | 0 | 1 | 2 | 3 => by decide
+  | n + 4 => by intros; linarith [n.add_div_right four_pos, n.div_le_self 4]
+
+/-- A list of Dihedral groups whose product will have commuting probability `1 / n`. -/
+def reciprocalFactors (n : ℕ) : List ℕ :=
+  if h0 : n = 0 then [0]
+  else if h1 : n = 1 then []
+  else if 2 ∣ n then
+    have := div_two_lt h0
+    3 :: reciprocalFactors (n / 2)
+  else
+    have := div_four_lt h0 h1
+    n % 4 * n :: reciprocalFactors (n / 4 + 1)
+
+@[simp] lemma reciprocalFactors_zero : reciprocalFactors 0 = [0] := rfl
+
+@[simp] lemma reciprocalFactors_one : reciprocalFactors 1 = [] := rfl
+
+lemma reciprocalFactors_def :
+    reciprocalFactors n =
+      if h0 : n = 0 then [0]
+      else if h1 : n = 1 then []
+      else if 2 ∣ n then
+        3 :: reciprocalFactors (n / 2)
+      else
+        n % 4 * n :: reciprocalFactors (n / 4 + 1) := by
+  conv_lhs => unfold reciprocalFactors
```

Comment:
By api lemmas, I meant something that hides the `if`s altogether
```suggestion
lemma reciprocalFactors_of_two_dvd (hn0 : n \ne 0) (hn : 2 | n) :
  reciprocalFactors n = 3 :: reciprocalFactors (n / 2) := sorry

lemma reciprocalFactors_of_not_two_dvd (hn0 : n \ne 0) (hn0 : n \ne 1) (hn : \not2 | n) :
  reciprocalFactors n = n % 4 * n :: reciprocalFactors (n / 4 + 1) := sorry
```
which should drop into your proof below without much work.

If you're feeling braver, *maybe* the following is a nice API:
```suggestion
lemma reciprocalFactors_two_mul (hn0 : n \ne 0) :
  reciprocalFactors (2*n) = 3 :: reciprocalFactors n := rfl
  
lemma reciprocalFactors_four_mul_add_one (hn0 : n \ne 0) :
  reciprocalFactors (4*n + 1) = (4*n + 1) :: reciprocalFactors (n + 1) := rfl
  
lemma reciprocalFactors_four_mul_add_three (hn0 : n \ne 0) :
  reciprocalFactors (4*n + 3) = 3*(4*n + 3) :: reciprocalFactors (n + 1) := rfl
```
if you think the